File: Progetto_1.py
# Metodi del Calcolo Scientifico
# Progetto_1 Python
# Mohammad Alì Manan (817205)
# Francesco Porto (816042)
# Stranieri Francesco (816551)

# https://docs.scipy.org/doc/scipy-0.14.0/reference/sparse.html
import scipy.io

# https://scikit-sparse.readthedocs.io/en/latest/cholmod.html
# https://readthedocs.org/projects/scikit-sparse/downloads/pdf/latest/
from sksparse.cholmod import cholesky

from numpy import ones, empty
from numpy.linalg import norm

import time
import os
import psutil

import pathlib
import platform as _platform
from datetime import datetime

import gc


def matrixSolver(matrix):
    fileNameMatrix = str(matrix).split(os.sep)[2]
    matrixName = fileNameMatrix.split(".")[0]
    print('MatrixName', matrixName)

    matrixProblem = scipy.io.loadmat(matrix)
    # print(sorted(matrixProblem.keys()))
    # print(matrixProblem['Problem'])

    # A = (matrixProblem['Problem'])['A'][0][0]

    matrixSize = ((matrixProblem['Problem'])['A'][0][0]).shape[0]
    print('MatrixSize', matrixSize)
    nnz = ((matrixProblem['Problem'])['A'][0][0]).getnnz()

    xEs = ones(matrixSize)
    b = ((matrixProblem['Problem'])['A'][0][0])*xEs
    x = empty(matrixSize)

    logFileWrite = True

    try:
        start = time.process_time()
        startMemoryAllocated = psutil.virtual_memory().used

        R = cholesky((matrixProblem['Problem'])['A'][0][0])
        x = R(b)

        endMemoryAllocated = psutil.virtual_memory().used
        # print((endMemoryAllocated.rss - startMemoryAllocated.rss) / 1000)

        executionTime = str(round((time.process_time() - start) * 1000))
        memoryAllocated = str(
            round((endMemoryAllocated - startMemoryAllocated) / 1000))
        relativeError = str("{:.7g}".format(norm(x - xEs) / norm(xEs)))
    except:
        executionTime = '0'
        memoryAllocated = '0'
        relativeError = '0'

        logFileWrite = False

    return (matrixName, matrixSize, nnz,
            executionTime, memoryAllocated, relativeError, logFileWrite)


def main():
    matrixFolder = "SuiteSparse/MAT/"
    matrixList = list(pathlib.Path(matrixFolder).glob('*.mat'))

    fileName = "result.txt"
    formatOut = "%d-%m-%Y %H:%M:%S"

    platform = _platform.system()
    if platform == 'Darwin':
        platform = 'MacOS'

    for matrix in matrixList:
        (matrixName, matrixSize, nnz,
         executionTime, memoryAllocated, relativeError, logFileWrite) = matrixSolver(matrix)

        gc.collect()

        print('ExecutionTime(ms)', executionTime)
        print('MemoryAllocated(KB)', memoryAllocated)
        print('RelativeError', relativeError)
        print()

        # logFile
        if logFileWrite:
            date = datetime.now().strftime(formatOut)

            with open(fileName, "a") as logFile:
                logFile.write('Date: ' + date + "\n" +
                              'Platform: ' + platform + "\n" +
                              'Language: ' + 'Python' + "\n" +
                              'MatrixName: ' + matrixName + "\n" +
                              'MatrixSize: ' + str(matrixSize) +
                              " x " + str(matrixSize) + "\n" +
                              'NonZero: ' + str(nnz) + "\n" +
                              'RelativeError: ' + relativeError + "\n" +
                              'ExecutionTime(ms): ' + executionTime + "\n" +
                              'MemoryAllocated(KB): ' + memoryAllocated + "\n" +
                              "\n")


if __name__ == "__main__":
    main()

File: Progetto_1.cpp
// Metodi del Calcolo Scientifico
// Progetto_1 C++
// Mohammad Alì Manan (817205)
// Francesco Porto (816042)
// Stranieri Francesco (816551)

#include <dirent.h>
#include <iostream>
#include <fstream>
#include <math.h>
#include <chrono>
#include <ctime>

#include "eigen-3.3.7/Eigen/Sparse"
#include "eigen-3.3.7/unsupported/Eigen/SparseExtra"

#if defined(_WIN32) || defined(__CYGWIN__)
#include <windows.h>
#include <psapi.h>
std::string platform("Windows");
#elif defined(__linux__) || defined(__linux) || defined(linux) || defined(__gnu_linux__)
#include <unistd.h>
#include <sys/resource.h>
std::string platform("Linux");
#elif defined(__APPLE__) && defined(__MACH__)
#include <mach/mach.h>
std::string platform("MacOS");
#else
#error Platform not supported!
#endif

// https://stackoverflow.com/questions/669438/how-to-get-memory-usage-at-runtime-using-c
/*
 * Author:  David Robert Nadeau
 * Site:    http://NadeauSoftware.com/
 * License: Creative Commons Attribution 3.0 Unported License
 *          http://creativecommons.org/licenses/by/3.0/deed.en_US
 */
size_t getCurrentRSS();

struct matrixSolved
{
        int matrixSize;
        int nnz;
        int memoryAllocated;
        int executionTime;
        double relativeError;
        bool logFileWrite;
};

matrixSolved matrixSolver(const char *matrixList);

using namespace Eigen;
using namespace std;
using namespace std::chrono;

typedef SparseMatrix<double> SpMat;

int main()
{
        DIR *matrixFolder;
        char *matrixName, *matrixExtension;
        struct dirent *dir;

        const char *matrixPath = "SuiteSparse/MTX/";
        matrixFolder = opendir(matrixPath);

        if (!matrixFolder)
        {
                cout << "Fatal Error! Folder NOT Found." << endl;
                return -1;
        }

        while ((dir = readdir(matrixFolder)) != NULL)
        {
                matrixName = strtok(dir->d_name, ".");
                matrixExtension = strtok(NULL, ".");

                if (matrixExtension == NULL || strcmp(matrixExtension, "mtx") != 0)
                        continue;

                int matrixListLength = sizeof(matrixPath) + sizeof(matrixName) + sizeof(matrixExtension) + 2;
                char matrixList[matrixListLength];

                strcpy(matrixList, matrixPath);
                strcat(matrixList, matrixName);
                strcat(matrixList, ".");
                strcat(matrixList, matrixExtension);

                cout << "MatrixName " << matrixName << endl;
                matrixSolved matrixSolved = matrixSolver(matrixList);

                cout << "ExecutionTime(ms) " << matrixSolved.executionTime << endl;
                cout << "MemoryAllocated(KB) " << matrixSolved.memoryAllocated << endl;
                cout << "RelativeError " << matrixSolved.relativeError << endl;

                // logFile
                if(matrixSolved.logFileWrite){
                        time_t now = system_clock::to_time_t(system_clock::now());
                        char date[20];
                        strftime(date, sizeof(date), "%d-%m-%Y %H:%M:%S", localtime(&now));

                        ofstream logFile;
                        logFile.open("result.txt", ios_base::app);
                        logFile << "Date: " << date << "\n";
                        logFile << "Platform: " << platform << "\n";
                        logFile << "Language: "
                                << "C++"
                                << "\n";
                        logFile << "MatrixName: " << matrixName << "\n";
                        logFile << "MatrixSize: " << matrixSolved.matrixSize << " x " << matrixSolved.matrixSize << "\n";
                        logFile << "NonZero: " << matrixSolved.nnz << "\n";
                        logFile << "RelativeError: " << matrixSolved.relativeError << "\n";
                        logFile << "ExecutionTime(ms): " << matrixSolved.executionTime << "\n";
                        logFile << "MemoryAllocated(KB): " << matrixSolved.memoryAllocated << "\n";
                        logFile
                        << "\n";
                        logFile.close();
                }
        }
        closedir(matrixFolder);
}

size_t getCurrentRSS()
{
#if defined(_WIN32) || defined(__CYGWIN__)
        PROCESS_MEMORY_COUNTERS info;
        GetProcessMemoryInfo(GetCurrentProcess(), &info, sizeof(info));
        return (size_t)info.WorkingSetSize;
#elif defined(__linux__) || defined(__linux) || defined(linux) || defined(__gnu_linux__)
        long rss = 0L;
        FILE *fp = NULL;
        if ((fp = fopen("/proc/self/statm", "r")) == NULL)
                return (size_t)0L; /* Can't open? */
        if (fscanf(fp, "%*s%ld", &rss) != 1)
        {
                fclose(fp);
                return (size_t)0L; /* Can't read? */
        }
        fclose(fp);
        return (size_t)rss * (size_t)sysconf(_SC_PAGESIZE);
#elif defined(__APPLE__) && defined(__MACH__)
        struct mach_task_basic_info info;
        mach_msg_type_number_t infoCount = MACH_TASK_BASIC_INFO_COUNT;
        if (task_info(mach_task_self(), MACH_TASK_BASIC_INFO,
                      (task_info_t)&info, &infoCount) != KERN_SUCCESS)
                return (size_t)0L; /* Can't access? */
        return (size_t)info.resident_size;
#else
#error Platform not supported!
        return (size_t)0L; /* Unsupported. */
#endif
}

matrixSolved matrixSolver(const char *matrixList)
{
        matrixSolved matrixSolved;

        // https://eigen.tuxfamily.org/dox/group__TutorialSparse.html
        SpMat A;
        loadMarket(A, matrixList);

        matrixSolved.matrixSize = A.rows();
        cout << "MatrixSize " << matrixSolved.matrixSize << endl;
        matrixSolved.nnz = A.nonZeros();

        VectorXd xEs = VectorXd::Ones(matrixSolved.matrixSize);
        VectorXd b = A * xEs;
        VectorXd x = VectorXd::Zero(matrixSolved.matrixSize);
        
        matrixSolved.logFileWrite = true;

        try
        {
                size_t memoryAllocatedStart = getCurrentRSS();
                high_resolution_clock::time_point start = high_resolution_clock::now();

                SimplicialLLT<SpMat> solver(A);
                x = solver.solve(b);

                high_resolution_clock::time_point stop = high_resolution_clock::now();
                duration<double, milli> differenceTime = (stop - start);
                size_t memoryAllocatedEnd = getCurrentRSS();

                matrixSolved.memoryAllocated = (memoryAllocatedEnd - memoryAllocatedStart) / 1000;
                matrixSolved.executionTime = round(differenceTime.count());
                matrixSolved.relativeError = (x - xEs).norm() / xEs.norm();
        }
        catch (...)
        {
                matrixSolved.memoryAllocated = 0;
                matrixSolved.executionTime = 0;
                matrixSolved.relativeError = 0;
                
                matrixSolved.logFileWrite = false;
        }

        return matrixSolved;
}
